### PR TITLE
Add a Javadoc section to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by pugnascotia**
Introduce a set of guidelines for when JavaDoc should be written, and when it shouldn't.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -237,9 +237,83 @@ Spotless will write files to
...
+      Elasticsearch repository, and it's easier to navigate when you
+      can quickly find out what is the purpose of a class. This doesn't
+      apply to inner classes or interfaces, unless you expect them to be
+      explicitly used outside their parent class.
+   3. New public methods must have JavaDoc, because they form part of the
+      contract between the class and its consumers. Similarly, new abstract
+      methods must have JavaDoc because they are part of the contract
+      between a class and its subclasses. It's important that contributors
+      know why they need to implement a method, and the JavaDoc should make
+      this clear.
```
Comment (not-napoleon) [inline CONTRIBUTING.md]: It's not clear to me from this section if a method which overrides a public abstract method should have javadoc or not, presuming the abstract method already does.  My gut instinct is "not unless it's doing something weird", but I'm open to other opinions.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -237,9 +237,83 @@ Spotless will write files to
...
-detect any misbehaviour. You can enabled the option from the command line
-by running Gradle with `-Dspotless.paddedcell`.
+The `paddedCell()` option is disabled for normal operation so that any
+misbehaviour is detected, and not just suppressed. You can enabled the
+option from the command line by running Gradle with `-Dspotless.paddedcell`.
+
+### JavaDoc
+
+Good JavaDoc can help with navigating and understanding code. Elasticsearch
+has some guidelines around when to write JavaDoc, and when not to.
```
Comment (jasontedor) [inline CONTRIBUTING.md]: Javadocs are good, but they’re narrow in purpose. They are suppose to say what a method does, its parameters and return values and exceptions, and some behavior details. One concern I have is that by being prescriptive about adding Javadocs, we’re potentially going to over-index on them, and neglect adding comments in method bodies to explain the *why* of a particular implementation (due to fatigue from writing Javadocs, etc.). I’d like our guidance to be accommodating for this.

**Review by jasontedor — COMMENTED** [review]
I left a comment.

**Comment by pugnascotia** [comment]
Thank you @not-napoleon and @jasontedor, I've tried to address your comments in 2c8919b.

**Review by jasontedor — COMMENTED** [review]
I like the change you pushed, thank you. One minor nit, they're called `Javadoc` these days (long ago they were called `JavaDoc`. Can you update that?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -231,15 +231,102 @@ from the command line.
...
+has some guidelines around when to write Javadoc and when not to, but note
+that we don't want to be overly prescriptive. The intent of these guidelines
+is to be helpful, not to turn writing code into a chore.
+
+#### The short version
+
+   1. Always add Javadoc to new code.
+   2. Add Javadoc to existing code if you can.
+   3. Document the "why", not the "how", unless that's important to the
+      "why.
```
Comment (matriv) [inline CONTRIBUTING.md]:
```suggestion
      "why".
```

**Review by matriv — COMMENTED** [review]
Thanks for doing that!
A couple of suggestions:

 - Maybe add a phrase for something obvious but still: The method/variable names should still be self-explanatory and descriptive and don't rely simply on the javadoc.
 - Maybe point out to use `{@link <class>/<method>}` for easy navigation also through the java docs.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Review by matriv — APPROVED** [review]
LGTM. Many thx again!

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Pull request merged**